src/plugins/discourse/references.js

```diff
@@ -23,3 +48,37 @@ export function parseLinks(cookedHtml: string): Hyperlink[] {
   parser.end();
   return links;
 }
+
+export function linksToReferences(
+  links: $ReadOnlyArray<Hyperlink>
+): DiscourseReference[] {
+  const server = "(https://[\\w.-]+)";
```

Comment:
This character class may require preprocessing. For example with `punycode` to allow http://見.香港/. Also missing the optional `https?`.

Reply:
Generally agree with deep then wide. Though I would consider this an accessibility issue, and correctness of the proposed feature, more than preemptively adding features. http://xn--nw2a.xn--j6w193g/ isn't legible as far as I'm aware. Though I'm ok with either choice.

Reply:
Based on a bit of research, it looks like Discourse already doesn't include non-punycode characters in the topic slug anyway. Take a look at this for an example: https://forums.eveonline.com/t/topic/195153

Going to go with "deep before wide" and merge as-is.

Reply:
I don’t think that it’s an accessibility issue. Such strings are simply
not valid URIs: [RFC 3986] says that characters must be “US-ASCII” and
defines the set of valid characters within a path component as

```
unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
```

and the [W3C URL spec] clearly specifies that each textual part of a URL
(e.g., the path, the query) is an ASCII string. Both of these standards
are referenced by the [HTML 5 spec].

Thus, if I write in a GitHub comment the Markdown

```markdown
[some link](http://見.香港/)
```

the rendered HTML that GitHub sends to the browser is actually

```html
<a href="http://%E8%A6%8B.%E9%A6%99%E6%B8%AF/" rel="nofollow">some link</a>
```

and, as @decentralion mentions, Discourse behaves similarly:

```html
<a href="http://xn--nw2a.xn--j6w193g/">some link</a>
```

This is because these servers are conforming to spec and sending valid
URIs. All that we need to do is properly parse percent signs (which we
might as well do). URIs are easy to parse _by design_; there is no need
to pull in Punycode conversions. To do otherwise wouldn’t be more
correct; it would be a non-standard extension.

There are lots of useful things to spend our limited energy on that will
probably have significantly higher impact. Let’s focus on those. :-)

[HTML 5 spec]: https://www.w3.org/TR/html50/infrastructure.html#valid-url
[RFC 3986]: https://tools.ietf.org/html/rfc3986#section-2
[W3C URL spec]: https://url.spec.whatwg.org/#concept-url

Reply:
Right, I was going to say we're not validating valid URIs but "cooked" html from a remote source. But if the HTML spec demands ASCII arguably it's the server doing it wrong if UTF stuff comes back from the server.